Implement enforcement of `trusted-types` CSP directive

lukewarlow opened the pull request:
#### 142d2a80207e10069b36196a19a83cda6b96223d
<pre>
Implement enforcement of `trusted-types` CSP directive
<a href="https://bugs.webkit.org/show_bug.cgi?id=267632">https://bugs.webkit.org/show_bug.cgi?id=267632</a>

Reviewed by Youenn Fablet.

This updates the trusted types policy creation code to validate that it&apos;s allowed by CSP.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicy-CSP-no-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-noNamesGiven-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-none-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-nameTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-list-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-list-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-reporting-check-report-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-reporting-expected.txt:
* Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
(WebCore::TrustedTypePolicyFactory::createPolicy):
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::allowTrustedTypesPolicy const):
(WebCore::ContentSecurityPolicy::reportViolation const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::violatedDirectiveForTrustedTypesPolicy const):
(WebCore::ContentSecurityPolicyDirectiveList::shouldReportSample const):
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.h:
* Source/WebCore/page/csp/ContentSecurityPolicyTrustedTypesDirective.cpp:
(WebCore::ContentSecurityPolicyTrustedTypesDirective::allows const):
* Source/WebCore/page/csp/ContentSecurityPolicyTrustedTypesDirective.h:

Canonical link: <a href="https://commits.webkit.org/274263@main">https://commits.webkit.org/274263@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90f30755aae96762685cde5c7fd0c03357c9516d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38339 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17282 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40666 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40896 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34095 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40586 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20048 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14620 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32325 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38912 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14589 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12693 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12671 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34268 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42175 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34887 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34678 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38517 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13259 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10968 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36718 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14820 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8643 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13666 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14284 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -45,8 +46,29 @@ Ref<TrustedTypePolicyFactory> TrustedTypePolicyFactory::create()
-ExceptionOr<Ref<TrustedTypePolicy>> TrustedTypePolicyFactory::createPolicy(ScriptExecutionContext&, const String& policyName, const TrustedTypePolicyOptions& options)
+ExceptionOr<Ref<TrustedTypePolicy>> TrustedTypePolicyFactory::createPolicy(ScriptExecutionContext& context, const String& policyName, const TrustedTypePolicyOptions& options)
 {
+    auto* csp = context.contentSecurityPolicy();
```
Could use CheckedRef

youennf commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -45,8 +46,29 @@ Ref<TrustedTypePolicyFactory> TrustedTypePolicyFactory::create()
+    auto* csp = context.contentSecurityPolicy();
+    ASSERT(csp);
+
+    AllowTrustedTypePolicyDetails details = { AllowTrustedTypePolicyDetails::Allowed };
```
```suggestion
    AllowTrustedTypePolicyDetails details = AllowTrustedTypePolicyDetails::Allowed;
```

youennf commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -754,6 +754,20 @@ bool ContentSecurityPolicy::allowBaseURI(const URL& url, bool overrideContentSec
     return allPoliciesAllow(WTFMove(handleViolatedDirective), &ContentSecurityPolicyDirectiveList::violatedDirectiveForBaseURI, url);
 }
 
+bool ContentSecurityPolicy::allowTrustedTypesPolicy(const String& value, bool isDuplicate, AllowTrustedTypePolicyDetails& details) const
```
I know this is the current style but it is not great to have both bool and the details our parameter.
I doubt we can have return value being false and details being Allowed.

Could we just have it as:
```suggestion
AllowTrustedTypePolicyDetails ContentSecurityPolicy::computeTrustedTypesPolicy(const String& value, bool isDuplicate) const
```

Then in call site, we could do a switch.

I would be tempted to rename AllowTrustedTypePolicyDetails to AllowTrustedTypePolicy

lukewarlow replied:
We can't have false and Allowed but we can have true and Disallowed because of report only mode. Though we only need access to details in this case within CSP itself not within the factory so I can probably refactor this like you suggest. Will give it a go.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-reporting-expected.txt:
```diff
@@ -1,12 +1,20 @@
+PASS Trusted Type violation report: creating a forbidden policy.
+PASS Trusted Type violation report: creating a report-only-forbidden policy.
+PASS Trusted Type violation report: creating a forbidden-but-not-reported policy.
+TIMEOUT Trusted Type violation report: assign string to script url Test timed out
```
This seems strange, can you check this?

If this is expected, we usually skip timing out tests until we fix the time out.

lukewarlow replied:
Just looked and this is expected, this PR only implements one of the two directives this test covers. Should I disable the test until both are done at the cost of less test coverage over this specific PR?

youennf replied:
Let's do that for now, are you planning to support it soon?
Otherwise, we could update the test to not timeout in the error case.

lukewarlow replied:
Yeah I plan on supporting that as a follow up PR (slightly larger so might take longer). Will disable this for now

youennf reviewed (commented):
LGTM but I have a question with one of the test result below.
Some additional suggestions below as well.

youennf commented on Source/WebCore/page/csp/ContentSecurityPolicyTrustedTypesDirective.h:
```diff
@@ -32,13 +32,13 @@
     ContentSecurityPolicyTrustedTypesDirective(const ContentSecurityPolicyDirectiveList&, const String& name, const String& value);
 
-    bool allows(const String& value, bool isDuplicate, AllowTrustedTypePolicyDetails&) const;
+    bool allows(const String& value, bool isDuplicate, AllowTrustedTypePolicy&) const;
```
It would be nice to change this one as well but probably not in this PR.

lukewarlow replied:
I'll leave it out of this PR but will revisit in future

lukewarlow committed:
Implement enforcement of `trusted-types` CSP directive
https://bugs.webkit.org/show_bug.cgi?id=267632

Reviewed by Youenn Fablet.

This updates the trusted types policy creation code to validate that it's allowed by CSP.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicy-CSP-no-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-noNamesGiven-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-none-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-nameTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-list-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-list-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-reporting-check-report-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-reporting-expected.txt:
* Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
(WebCore::TrustedTypePolicyFactory::createPolicy):
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::allowTrustedTypesPolicy const):
(WebCore::ContentSecurityPolicy::reportViolation const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::violatedDirectiveForTrustedTypesPolicy const):
(WebCore::ContentSecurityPolicyDirectiveList::shouldReportSample const):
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.h:
* Source/WebCore/page/csp/ContentSecurityPolicyTrustedTypesDirective.cpp:
(WebCore::ContentSecurityPolicyTrustedTypesDirective::allows const):
* Source/WebCore/page/csp/ContentSecurityPolicyTrustedTypesDirective.h:

Canonical link: https://commits.webkit.org/274263@main
